inference-engine/src/transformations/src/transformations/common_optimizations/mul_fake_quantize_fusion.cpp

```diff
@@ -0,0 +1,106 @@
+                                                                        ngraph::pattern::any_input(),
+                                                                        ngraph::pattern::any_input()});
+    ngraph::matcher_pass_callback callback = [=](pattern::Matcher& m) {
+        auto fq = std::dynamic_pointer_cast<opset5::FakeQuantize>(m.get_match_root());
+        if (!fq)
+            return false;
+        auto mul = std::dynamic_pointer_cast<opset5::Multiply>(fq->input_value(0).get_node_shared_ptr());
+        if (!mul)
+            return false;
+        auto mul_data = mul->input_value(0).get_node_shared_ptr();
+        auto mul_2nd_input = mul->input_value(1).get_node_shared_ptr();
+        auto mul_const = std::dynamic_pointer_cast<opset5::Constant>(mul_2nd_input);
+        if (!mul_const) {
+            return false;
+        }
+
+        auto shape = mul_const->get_shape();
+        size_t const_shape_size = shape_size(shape);
+        if (const_shape_size > 1) {
+            // disallow constant shapes other than (N, 1, 1, ..., 1) or (1, C, 1, ..., 1)
+            if (!(shape[0] > 1 && shape[0] == const_shape_size) &&
+                !(shape.size() > 1 && shape[1] == const_shape_size)) {
+                return false;
+            }
+        }
+
+        if (const_shape_size > 1 &&
+            static_cast<Dimension::value_type>(shape.size()) < fq->get_input_partial_shape(0).rank().get_length()) {
+            // Reshape constants like (C, 1, 1) to (1, C, 1, 1)
+            shape.insert(shape.begin(), fq->get_input_partial_shape(0).rank().get_length() - shape.size(), 1);
+            mul_2nd_input = std::make_shared<opset5::Reshape>(mul_2nd_input, op::Constant::create(element::u64, Shape{shape.size()}, shape), false);
+        }
+
+        auto new_input_low = std::make_shared<opset5::Divide>(fq->input_value(1), mul_2nd_input);
+        auto new_input_high = std::make_shared<opset5::Divide>(fq->input_value(2), mul_2nd_input);
+        auto new_output_low = fq->input_value(3).get_node_shared_ptr();
+        auto new_output_high = fq->input_value(4).get_node_shared_ptr();
+
+        std::shared_ptr<Node> new_fq;
+        auto mul_const_value = mul_const->cast_vector<float>();
+        if (std::all_of(mul_const_value.begin(), mul_const_value.end(), [] (float f) -> bool { return f < 0.0f; })) {
+            new_output_low = fq->input_value(4).get_node_shared_ptr();
+            new_output_high = fq->input_value(3).get_node_shared_ptr();
+            new_fq = register_new_node<opset5::FakeQuantize>(mul_data, new_input_low, new_input_high,
+                    fq->input_value(4), fq->input_value(3), fq->get_levels());
+            copy_runtime_info({mul, fq}, {mul_2nd_input, new_input_low, new_input_high, new_fq});
+        } else if (std::any_of(mul_const_value.begin(), mul_const_value.end(), [] (float f) -> bool { return f < 0.0f; })) {
```

Comment:
Hard to read without comments. Please leave a scheme of what's going on.

Reply:
Did you see the code in MO https://github.com/openvinotoolkit/openvino/blob/master/model-optimizer/extensions/middle/MulFakeQuantizeFuse.py
It has much simpler logic. Why wasn't it reused. Is there smth wrong with the logic there?

Reply:
There is nothing wrong with the logic in MO, I just didn't want to operate on raw data from constants, because I'd have to implement broadcasting there and I was reluctant to do this. Anyway, I've simplified that a little and put some comments there